meteor-bundle-main.js

```diff
@@ -73,29 +113,150 @@ function monkeypatchHttpAndHttps() {
             continue;
           } else {
-            throw e;
+            // Sometimes the server gives us a 0-byte response,
+            // presumably due to a connection getting reset?.
+            console.error("Got exception reading JSON from file:", metadataFilename, "Dazed and confused, but trying to continue.", e);
+            continue;
           }
         }
 
-        // If this certificate isn't valid yet, keep looping, hoping
-        // to find one that is valid.
-        var notBefore = metadata['notBefore'];
-        if (notBefore && (notBefore < nowUnixTimestamp)) {
+        // If the cert is expired, definitely don't use it.
+        if (now > metadata.notAfter) {
+          console.log("Skipping", keyFilename, "because", metadata.notAfter, "is in the past.");
           continue;
         }
 
-        result['ca'] = metadata['ca'];
-        result['key'] = fs.readFileSync(keyFilename, 'utf-8');
-        result['cert'] = metadata['cert'];
-        return result;
+        validCertificates.push(metadata);
+      }
+
+      // Sort by notBefore, ascending.
+      validCertificates.sort(function(cert1, cert2) {
+        return (cert1.notBefore - cert2.notBefore);
+      });
+
+      if (! validCertificates) {
+        // TODO: Be more robust here, though it's not clear how.
+        console.error("Aieee, we found zero valid certificates. Allowing Sandstorm to crash.");
+        return;
       }
+
+      console.log("Using this HTTPS key:", validCertificates[0].keyFilename,
+                  "valid starting", new Date(validCertificates[0].notBefore),
+                  "until", new Date(validCertificates[0].notAfter));
+
+      // Store HTTPS configuration data from the oldest certificate.
+      var result = {
+        ca: validCertificates[0].ca,
+        key: fs.readFileSync(validCertificates[0].keyFilename, 'utf-8'),
+        cert: validCertificates[0].cert
+      };
+
+      // Calculate re-key time.
+      //
+      // - If the cert we want to switch to is not valid yet, then we
+      //   re-key at its notBefore time.
+      //
+      // - If the cert we want to switch to *is* already valid, then
+      //   we re-key at the notAfter time of our current cert.
+      //
+      // - If there is no cert we want to switch to, then the
+      //   nextRekeyTime is null.
+      result.nextRekeyTime = null;
+
+      if (validCertificates.length >= 2) {
+        var secondBest = validCertificates[1];
+        if (now < secondBest.notBefore) {
+          result.nextRekeyTime = secondBest.notBefore;
+        } else {
+          result.nextRekeyTime = validCertificates[0].notAfter;
+        }
+
+        console.log("Will switch to", validCertificates[1].keyFilename,
+                    "at time", new Date(result.nextRekeyTime),
+                    "via real-time SNI re-keying system.");
+      } else {
+        console.log("We have only the one key that we know about. We'll need to renew soon.");
```

Comment:
The current strategy here is to have two certificates ready at all times. This is for two reasons: (1) It made implementing things easier and (2) if the Sandcats service is (somehow) down for a while, or super buggy, or people turn off their machine and then turn it back on a week later or something like that, I don't want people's certs to expire. (Of course, it _will_ auto re-key on startup, so maybe I could worry about that less.)

Alternate strategy 1: If this is your first cert, get just 1 week, after that, get 2 weeks

* Day 0: new install: have a duration of 1 week (first cert)

* Day 6: Fetch the next week (second cert)

* Day 7: Switch to second cert

* Day 7: Fetch a week after that (third cert)

Alternate strategy 2:

* Always do the renewal one day before the cert is going to expire. This scares me because it gives us a smaller window to deal with sandcats service bugs.

@kentonv do you have a preference for "Alternate strategy 1" or "Alternate strategy 2"?